add test code

Ikumi-Kobayashi opened the pull request:
テストコードを作成しました。

Commit:
add test code

shnzw124 commented on app/services/poker_check_service.rb:
```diff
@@ -46,8 +46,12 @@ def error
         #カードの重複
-       else  @card.count - @card.uniq.count > 0
-         @error_message = "カードが重複しています。"
+          if @card[0] == @card[1] or @card[0] == @card[2] or @card[0] == @card[3] or @card[0] == @card[4] or
```
配列内の要素の重複をチェックするメソッドを利用すればここの行は一行で書き換えれます。
「Ruby 配列 重複」などで調べてみましょう

shnzw124 commented on spec/services/poker_check_service_spec.rb:
```diff
@@ -0,0 +1,110 @@
+#こんな感じでうまく動いてるから、後はどんな例で試すか決めて書いていくだけ！
+RSpec.describe PokerCheckService do
+  describe '#error' do
+    it 'カードが５個じゃないとき' do
```
Rspecのテストコードは
自動テストであると同時に アプリケーションの仕様です。
※仕様とは？...https://wa3.i-3-i.info/word1942.html

it 'カードが５個じゃないとき'
が3つ存在してますが、それぞれどういう内容でテストしてるのかをもうちょっと詳しく書きましょう。
（この場合だと手札が0枚のとき、など）

shnzw124 commented on spec/services/poker_check_service_spec.rb:
```diff
@@ -0,0 +1,110 @@
+require 'rails_helper'
```
Service Specのテストケースとしては網羅できているのでOKです！


Controllerのテストコードが実装されていないようなので
そちらも追加しましょう

以下のページを参考にして
MVCモデルにおけるControllerの役割とController Specで確認すべきテストケースをインプットしてください↓

https://forest-valley17.hatenablog.com/entry/2018/10/15/140222
https://blog.naichilab.com/entry/2016/01/19/011514#コントローラの責務
https://qiita.com/t2kojima/items/ad7a8ade9e7a99fb4384

shnzw124 commented on spec/services/poker_check_service_spec.rb:
```diff
@@ -0,0 +1,110 @@
+RSpec.describe PokerCheckService do
+  describe '#error' do
+    it 'カードが５個じゃないとき' do
+      poker_check_service = PokerCheckService::PokerCheckService.new("")
```
このファイル内でPokerCheckServiceをincludeしておけば
PokerCheckService.new("")で記載できるはずです

Ikumi-Kobayashi committed:
重複の書き方を一行にしました。

Ikumi-Kobayashi committed:
重複の書き方を一行にしました。
３つあった　it'カードが５個じゃないとき' をもっと具体的にしました。

Ikumi-Kobayashi committed:
重複の書き方を一行にしました。
３つあった　it'カードが５個じゃないとき' をもっと具体的にしました。
controllerのテストも書いたのですが、うまくいきません。レビューお願いします。

Ikumi-Kobayashi committed:
重複の書き方を一行にしました。
３つあった　it'カードが５個じゃないとき' をもっと具体的にしました。
controllerのテストも書いたのですが、うまくいきません。（poker_controller_spec.rbの11~19行目のメモにしてあるところ）レビューお願いします。

Ikumi-Kobayashi committed:
重複の書き方を一行にしました。
３つあった　it'カードが５個じゃないとき' をもっと具体的にしました。
controllerのテストも書いたのですが、うまくいきません。（poker_controller_spec.rbの11~19行目のメモにしてあるところ）レビューお願いします。
include PokerCheckServiceをpoker_check_service_spec.rbでしても、PokerCheckService.new("")で記載するとエラーになってしまったので、そのままの書き方で置いてあります。

Ikumi-Kobayashi committed:
test modified

shnzw124 commented on spec/controllers/poker_controller_spec.rb:
```diff
@@ -0,0 +1,21 @@
+  include PokerCheckService
+
+RSpec.describe PokerController, type: :controller do
+  describe 'controller全般' do
```
リクエスト成功時はステータスコード200で返しますが
リクエスト失敗時は失敗内容によって400台のステータスコードを返却します。
そちらの確認もお願いします。

Ikumi-Kobayashi committed:
make API code

shnzw124 commented on app/api/v1/check.rb:
```diff
@@ -0,0 +1,88 @@
+module V1
```
api名は○○_api.rbとしましょう。

shnzw124 commented on app/api/v1/check.rb:
```diff
@@ -0,0 +1,88 @@
+module V1
+  class Check < Grape::API
+
+    include PokerCheckServiceApi
```
今回API用に新しくServiceクラスをつくってる？
Webアプリ作成時につくったServiceクラスの中に必要なメソッドを追記する形で実装したほうがよいです。

同じようなコードをいくつものファイルに書いていると
保守（＝プログラム修正）の際に大変なので
できるだけ共通の処理はまとめましょう

shnzw124 commented on app/api/v1/check.rb:
```diff
@@ -0,0 +1,88 @@
+
+    resources :check do
+      post '/' do
+        api_result = {}
```
変数は最初にいっきに定義してしまうするよりも
実際に使う直前に定義するほうがよいです

変数を定義した位置と実際に使う位置が離れていると
どういった変数だったか捉えづらくなるので

shnzw124 commented on spec/services/poker_check_service_api_spec.rb:
```diff
@@ -0,0 +1,111 @@
+require 'rails_helper'
```
今回新たにAPI用にServiceを実装してもらっていますが不要です。
既存のServiceクラスにAPIで必要なメソッドを追記して、それに対応するService Specに
テストコードを追記してください。

また、Service Specへの追記と併せて
APIのRequest Specを作成してください。
Request SpecはWebアプリでいうController Specと同様のものです。

shnzw124 commented on app/api/v1/check.rb:
```diff
@@ -0,0 +1,88 @@
+              best = "false"
+            end
+            #エラーだけの時
+            if error_message.present? && result.blank?
```
Grape::Entityというものを使うとより綺麗に書けるとおもいます。
https://github.com/ruby-grape/grape-entity

CardEntityとErrorEntityとResponseEntityを定義して
CardEntityにはレスポンスとしてcard/hand/best、ErrorEntityにはcard/msgを含めるとして
ResponseEntityにCardEntityとErrorEntityを利用する
みたいに書くといけるはず

shnzw124 reviewed (commented):
コメントしました。

綺麗で読みやすいコードを書くために、以下の記事読んで見るといいいですよ！
https://qiita.com/AKB428/items/20e81ccc8d9998b5535d

「リーダブルコード」はぜひ読んでみてください

Ikumi-Kobayashi committed:
apiのファイル名をcheck_api.rb、root_api.rbに変更しました。

Ikumi-Kobayashi committed:
apiのファイル名をcheck_api.rb、root_api.rbに変更しました。
新たに作ったサービスを削除し、既存のサービスにstrongアクションを追加する形にしました。

Ikumi-Kobayashi committed:
apiのファイル名をcheck_api.rb、root_api.rbに変更しました。
新たに作ったサービスを削除し、既存のサービスにstrongアクションを追加する形にしました。
api_resultの変数定義の場所を変更しました。

Ikumi-Kobayashi committed:
apiのファイル名をcheck_api.rb、root_api.rbに変更しました。
新たに作ったサービスを削除し、既存のサービスにstrongアクションを追加する形にしました。
api_resultの変数定義の場所を変更しました。
zipメソッドでまとめて処理していたところを、仏様な処理ごとに分割しました。

Ikumi-Kobayashi committed:
apiのファイル名をcheck_api.rb、root_api.rbに変更しました。
新たに作ったサービスを削除し、既存のサービスにstrongアクションを追加する形にしました。
api_resultの変数定義の場所を変更しました。
zipメソッドでまとめて処理していたところを、必要な処理ごとに分割しました。
entityを用いて処理を書き直しました。

Ikumi-Kobayashi committed:
modified API code

Ikumi-Kobayashi committed:
modified API code

Ikumi-Kobayashi committed:
課題２および課題２のテストコードを作成しました。
